source/common/buffer/buffer_impl.cc

```diff
@@ -44,171 +62,574 @@ void OwnedImpl::add(const Instance& data) {
+      ASSERT(dest == static_cast<const uint8_t*>(reservation.mem_) + linearized_size);
+      new_slice->commit(reservation);
+      slices_.emplace_front(std::move(new_slice));
+    }
+    return slices_.front()->data();
+  }
 }
 
 void OwnedImpl::move(Instance& rhs) {
   ASSERT(&rhs != this);
-  // We do the static cast here because in practice we only have one buffer implementation right
-  // now and this is safe. Using the evbuffer move routines require having access to both evbuffers.
-  // This is a reasonable compromise in a high performance path where we want to maintain an
-  // abstraction in case we get rid of evbuffer later.
-  int rc = evbuffer_add_buffer(buffer_.get(), static_cast<LibEventInstance&>(rhs).buffer().get());
-  ASSERT(rc == 0);
-  static_cast<LibEventInstance&>(rhs).postProcess();
+  ASSERT(isSameBufferImpl(rhs));
+  if (old_impl_) {
+    // We do the static cast here because in practice we only have one buffer implementation right
+    // now and this is safe. Using the evbuffer move routines require having access to both
+    // evbuffers. This is a reasonable compromise in a high performance path where we want to
+    // maintain an abstraction in case we get rid of evbuffer later.
+    ASSERT(dynamic_cast<LibEventInstance*>(&rhs) != nullptr);
+    int rc = evbuffer_add_buffer(buffer_.get(), static_cast<LibEventInstance&>(rhs).buffer().get());
+    ASSERT(rc == 0);
+    static_cast<LibEventInstance&>(rhs).postProcess();
+  } else {
+    // We do the static cast here because in practice we only have one buffer implementation right
+    // now and this is safe. This is a reasonable compromise in a high performance path where we
+    // want to maintain an abstraction.
+    ASSERT(dynamic_cast<OwnedImpl*>(&rhs) != nullptr);
+    OwnedImpl& other = static_cast<OwnedImpl&>(rhs);
+    while (!other.slices_.empty()) {
+      const uint64_t slice_size = other.slices_.front()->dataSize();
+      slices_.emplace_back(std::move(other.slices_.front()));
+      other.slices_.pop_front();
+      length_ += slice_size;
+      other.length_ -= slice_size;
+    }
+    other.postProcess();
+  }
 }
 
 void OwnedImpl::move(Instance& rhs, uint64_t length) {
   ASSERT(&rhs != this);
-  // See move() above for why we do the static cast.
-  int rc = evbuffer_remove_buffer(static_cast<LibEventInstance&>(rhs).buffer().get(), buffer_.get(),
-                                  length);
-  ASSERT(static_cast<uint64_t>(rc) == length);
-  static_cast<LibEventInstance&>(rhs).postProcess();
+  ASSERT(isSameBufferImpl(rhs));
+  if (old_impl_) {
+    // See move() above for why we do the static cast.
+    int rc = evbuffer_remove_buffer(static_cast<LibEventInstance&>(rhs).buffer().get(),
+                                    buffer_.get(), length);
+    ASSERT(static_cast<uint64_t>(rc) == length);
+    static_cast<LibEventInstance&>(rhs).postProcess();
+  } else {
+    // See move() above for why we do the static cast.
```

Comment:
Now that you mention it, the dynamic cast is redundant now, because the `ASSERT(isSameBufferImpl(rhs))` at the start of the method already encapsulates the same check (and does additional, more thorough checks). I'll remove the dynamic_cast here.